File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<string> dfs(const vector<vector<string>>& tickets, vector<bool>& used, string cur, vector<string>& path) {
    if (find(used.begin(), used.end(), false) == used.end()) return path;
    
    for (int i=0; i<tickets.size(); i++) {
        if (tickets[i][0] == cur && !used[i]) {
            used[i] = true;
            path.push_back(tickets[i][1]);
            
            vector<string> res = dfs(tickets, used, tickets[i][1], path);
            if (!res.empty()) return res;
            
            used[i] = false;
            path.pop_back();
        }
    }
    
    return {};
}

vector<string> solution(vector<vector<string>> tickets) {
    sort(tickets.begin(), tickets.end(), [](const vector<string>& a, const vector<string>& b) {return a[1] < b[1];});
    
    int n = tickets.size();
    vector<bool> used(n, false);
    vector<string> path = {"ICN"};
    
    return dfs(tickets, used, "ICN", path);
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    
    def dfs(cur, path):
        if len(path) == len(tickets) + 1:
            return path
        
        # 다음 목적지 찾아 가기
        for i, (depart, dest) in enumerate(tickets):
            if depart == cur and not visited[i]:
                visited[i] = True
                result = dfs(dest, path + [dest])
                
                if result: # 정답 찾았으면 전달
                    return result
                
                visited[i] = False
        return None
                
    # 목적지를 알파벳 순서대로 정렬
    tickets.sort(key=lambda x: x[1])
    visited = [False] * len(tickets)
    
    return dfs("ICN", ["ICN"])